toto/verifylib.py

```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
+
+  path = rule[1]
+  if (path not in link.materials.keys()):
+    raise RuleVerficationFailed("'%s' " \
+        "not found in materials of link '%s'" % (path, link.name))
+
+  if (path in link.products.keys()):
+    raise RuleVerficationFailed("'%s' found in products of link '%s' "
+        "- should have been deleted" % (path, link.name))
+
+
+def verify_modify_rule(rule, link):
+  """
+  <Purpose>
+    Verifies that the path (2nd element in rule list) is found in the materials
+    and products list of the passed Link object and that the hashes of the
+    according material and product are not equal, i.e. the file was modified in
+    the step the rule was defined for.
+
+  <Arguments>
+    rule:
+            The rule to be verified. Format is: ["MODIFY", "<path>"]
+
+    link:
+            The Link object for the Item (Step or Inspection) that contains
+            the rule.
+
+  <Exceptions>
+    raises an Exception if the rule is not conformant with the rule format.
+    raises an if a matchrule does not verify.
+    TBA (see https://github.com/in-toto/in-toto/issues/6)
+
+    RuleVerficationFailed
+
+  <Side Effects>
+    None.
+
+  """
+  # FIXME: Validate rule format
+
+  path = rule[1]
+
+  if (path not in link.materials.keys()):
+    raise RuleVerficationFailed("'%s' " \
+        "not found in materials of link '%s'" % (path, link.name))
+
+  if (path not in item_link.products.keys()):
+    raise RuleVerficationFailed("'%s' "
+        "not found in products of link '%s'" % (path, link.name))
+
+  if (ComparableHashDict(link.materials[path]) == \
+      ComparableHashDict(link.products[path])):
+    raise RuleVerficationFailed("hashes of product and material '%s' of link "
+        "'%s' match - should have been modified" % (path, link.name))
+
+
+def _verify_rules(rules, source_type, source_link, target_links):
+  """
+  <Purpose>
+    Helper method to iteratively verify all rules of a type.
+
+  <Arguments>
+    rules:
+            A list containing rules defined in the material_matchrules or
+            product_matchrules field of a Step or Inspection object.
+
+    source_type:
+            A string to identify if the rule is a material matchrule or
+            product matchrule. One of "material" or "product".
+
+    source_link:
+            The Link object for an Item (Step or Inspection) that contains the
+            rules to be verified.
+            The contained materials and products are used as verification source.
+
+    target_links:
+            A dictionary of Link objects with Link names as keys.
+            The Link objects relate to Steps.
+            The contained materials and products are used as verification target.
+
+  <Exceptions>
+    raises an Exception if a rule is not conformant with the rule format.
+    raises an Exception if a matchrule does not verify.
+    TBA (see https://github.com/in-toto/in-toto/issues/6)
+
+  <Side Effects>
+    Calls the respective verify_<rule-type>_rule function.
```

Comment:
You are right. I don't think it is a side-effect either. I guess I saw some similar "side effects" in the tuf projects and adapted the broad understanding. I'll take this out here.